### DISCUSSION TITLE
Release

### DISCUSSION THREAD

**Discussion by ballsteve** [ballsteve/xrust#19] [General]
I'm getting the package ready for another release.

I've tidied up the code to eliminate all rustc warnings. We'll also run all tests and make sure they all pass.

Once that's done I'll push out v0.6.

**Comment by ballsteve** [comment]
Oh, and merge the dev branch in to main.

**Comment by ballsteve** [comment]
Oh, and eliminate all panics from the library.

**Reply by ballsteve** [reply]
I've eliminated all of the easy panics (replaced .expect("") with ?).

Rust doesn't like ? inside a closure. A couple of those instances have been fixed, but it is not straight-forward to do that.

I've also found a few instances where the function API has to change, i.e. where a function returns bool it must instead return Result<bool, Error>, so the caller has to handle the Result as well. This makes the change more complicated.

I'm putting the complete elimination of panics in the too hard basket, i.e. making it an issue.

**Comment by Devasta** [comment]
Hi Steve,

Just sent a PR in there, 

conformance::xml::xmltest_valid_sa_canonicalonly::validsa052
conformance::xml::xmltest_valid_sa_canonicalonly::validsa064
conformance::xml::xmltest_valid_sa_canonicalonly::validsa089

I still need to sort these three unit tests and then I should be good, will hopefully have something for you in next couple of days..

**Comment by Devasta** [comment]
Actually, nevermind, those are fixed now!

test result: ok. 544 passed; 0 failed; 234 ignored; 0 measured; 0 filtered out; finished in 0.26s

I think I've gotten the tests about as far as they'll go without DTDs. I don't use them at all at work, so I'll need to a bit of reading of the spec first but I will see whats required...

**Comment by ballsteve** [comment]
dev is now compiling without warnings and all tests pass.

I'll promote the code to the main branch and push out the next release: 0.6.0

**Comment by ballsteve** [comment]
Release 0.6.0 is done!

Published on crates.io

Thank you Daniel for your efforts to get the XML parser into a much better state!

**Reply by Devasta** [reply]
No problem! 😄

**Comment by ballsteve** [comment]
I'm working on xsl:include and xsl:import now.

**Comment by ballsteve** [comment]
Now planning release 0.7.0. At this stage it will include:

* Forest module
* xsl:include, xsl:import
* general entity expansion

**Comment by ballsteve** [comment]
I'm thinking of experimenting with string interning too.

**Comment by ballsteve** [comment]
A new Pull Request has been created for the next release.

Please comment on this PR, and once we're happy with it we'll push out the next release.

**Comment by ballsteve** [comment]
The PR has been merged.

I'm going to finalise the documentation, and then make a new release.

**Comment by ballsteve** [comment]
The dev branch is now passing all tests, but is producing a lot of warnings.

Once the warnings have been dealt with, we'll then release. This will be the "parser combinator" release.

After that, I'll concentrate on getting the next release with the transformation combinator.

**Reply by Devasta** [reply]
I've not run my changes in Dev through clippy yet, I'll try to get to that this week.

**Reply by Devasta** [reply]
Right, sent another PR in with a bunch of clippy fixes. I just did the parser, but I should have time Friday to sweep through the whole thing and cargo-fmt the project if you like?

**Reply by ballsteve** [reply]
That sounds great! I've done a few fixes, so by the end of the week we should be able to merge it all together and push out a release.

**Reply by ballsteve** [reply]
I've done some cleanup and run cargo fmt.

The only thing left to do (from my pov) is to make the function that resolves a URL for include/import accept a closure.

**Reply by Devasta** [reply]
I've done some more cleanup, one clippy warning left (that I am not sure how to fix) and the unit tests need to have old dependencies updated.

We're close! 😄

**Reply by Devasta** [reply]
And I just did a review of the outstanding conformance tests, another 100 passing.

**Reply by ballsteve** [reply]
Woo-hoo!

I'll review the documentation, and then get the release done

**Reply by ballsteve** [reply]
So that its documentation can be included, I've changed the parser module to be public. However, all of the functions are pub(crate) so nothing will be public anyway. Did you intend the new parser combinator to be entirely private, or would you like to open it up for others to use it?

**Reply by Devasta** [reply]
We can make it public, thats no problem. I wasn't exactly sure what the scope for the project would be, like if we were "just" going focus on XSLT or would we expose the XML reader/writer and XDM for other applications to use.

**Reply by ballsteve** [reply]
Not everyone/everything uses XSLT. There are plenty of useful things to do with XPath, trees (XDM), evaluation, and parsing. For example, it wouldn't be too hard to adapt the evaluation engine to do XQuery. Another follow-on project of mine is using Xrust in WASM; XSLT doesn't really make that much sense in that environment.

If you're OK with it, I'll leave the parser combinator as public. I've added some placeholder documentation. That can be fleshed out in a later release.

**Reply by Devasta** [reply]
Not everyone uses XSLT... But they should! 😁

Sounds good, I'll flesh out some documentation when I can.

**Comment by ballsteve** [comment]
I've published xrust 0.8 to crates.io, but I jumped the gun and did it from the dev branch (oops!).

Let's get everything merged into the main branch (PR now pending), and then move on with the next stage of development

**Reply by Devasta** [reply]
Approved the PR, I'll get it merged when I get home.

**Reply by Devasta** [reply]
And we're live! :D

**Reply by ballsteve** [reply]
Thanks mate!

My plan for 0.9 is to:

1. Cleanup unused/obsolete modules
2. Complete the transformation combinator
3. Improve the tree module by separating the node structure from node content. My intention is to make support for XML features better, such as XML Declaration, DTDs, etc.
4. Write a Markdown parser (for the Content Management System that's on my wish-list)

**Reply by Devasta** [reply]
I'd like to start building external DTD support, I just need to look at library choices for pulling from the web. I figure whatever we use we would also want for xsl:include and xsl:import?

I was also thinking, if the transformation Combinators work well that we could look at doing something similar for document validation.

All that said, I'm happy to work on whatever you think would be best.

**Reply by ballsteve** [reply]
What I found is that including networking code in the crate breaks Wasm. To fix this, I've split the include/import function into a separate crate (xrust-net). In the main xrust crate, there is a closure that resolves a URL to the content of the resource, see xslt.rs.

That way, a Wasm program can not import the xrust-net crate and instead provide its own mechanism to resolve URLs, presumably using the browser's facilities via JavaScript.

I'll upload the xrust-net crate source code to the Github repo. I haven't yet published it as a separate crate.

Cheers, Steve

**Reply by ballsteve** [reply]
xrust-net is now up on GitHub in my personal repo.

**Reply by Devasta** [reply]
Hi Steve, looking at xrust-net, I can see it has xrust as a dependency... should that be the other way around, having it as an optional dependency of xrust?

**Reply by ballsteve** [reply]
The aim is to remove dependencies on network and filesystem libraries from the main xrust crate. That way it can be used with wasm. There is probably a better way to do that, but I haven't found it yet.

I tried making the network parts of xrust an optional dependency, but it didn't work. I could make the compilation part optional, but not the cargo dependency.

**Comment by ballsteve** [comment]
It's time for another release!

I've created a PR for my changes to the transformation module. Since I couldn't get the combinator idea working, I've created a Transform module that is a data structure to describe the transformation, an interpreter, and parsers for XPath, Patterns and XSLT.

I want to merge that work into the main branch, along with Daniel's work on the parser combinator, and do some cleanup. For the latter, I'm planning on creating a 'Trees' module and putting the various tree implementations in that.

It would also be nice to have a few new features. I'm thinking of adding a Markdown parser. Any other ideas?

**Reply by ballsteve** [reply]
Regarding cleanup, we also need to get all examples working and update/improve documentation.

**Reply by Devasta** [reply]
Invisible XML would be a very good spec to implement if we can.

I'd like to work a bit more on building out validation if possible, but I'm a little out of my depth. 😅

**Reply by ballsteve** [reply]
I looked at validation a few years ago, and came to the same conclusion ;-)

I assume you're talking DTD validation, but I would consider that to be a subset of XSD, or even RelaxNG, validation.

Perhaps the libxml2 library could provide insight/inspiration?

**Reply by ballsteve** [reply]
There's already a Rust implementation of ixml - EarleyBird. Perhaps we could use that?

**Reply by Devasta** [reply]
Tbh if we supports XML IDs and had a catalogue for common DTDs for entity expansion that'd be good enough for DTDs I reckon, especially since their interactions with namespaces arent great.

XSD support would be incredible if we managed it.

**Reply by Devasta** [reply]
Definitely worth a look at earleybird, I'll try and check it out this weekend if I get a chance (I'm currently looking at wedding venues!)

**Reply by ballsteve** [reply]
Well, that sounds much more important! Congrats on your upcoming nuptials!

**Reply by ballsteve** [reply]
The dev branch is now a release candidate. It passes all tests.

The only thing left to do now is update documentation, merge into main, and then publish.

**Reply by Devasta** [reply]
Looking forward to its release. 😄

**Reply by ballsteve** [reply]
Thanks for approving the PR.

I've removed an obsolete test script, as well as examples/issue_52.rs since it creates a dependency on network I/O that we're trying to avoid for Wasm.

All tests pass, so we're ready for release.

**Reply by ballsteve** [reply]
It is done!

Release 0.9.0 of 𝛘rust is now published on crates.io.

**Comment by ballsteve** [comment]
# Version 1.0

What's left to do for Version 1.0? That is, what features are there in XSLT 1.0 for us to implement. Here they are:

* modes
* sorting
* ---copy-of--- (already implemented)
* document()
* format-number()
* xsl:decimal-format
* xsl:number
* attributes sets
* extensions
* abbreviated syntax for XPath

**Reply by ballsteve** [reply]
I forgot to add "abbreviated syntax". I've been hoping that someone would contribute that as their own sub-project, but it hasn't happened so I guess I'll have to do it :-(

**Reply by ballsteve** [reply]
modal processing is now in the may-2024 branch.

Sorting shouldn't be too hard, so I'll tackle that next. I'm open to suggestions...

**Reply by ballsteve** [reply]
Sorting is now done. Checked into may-2024 branch.

**Reply by ballsteve** [reply]
format-number is done and in the may-2024 branch. However, I've decided that decimal-format is not a high priority at this stage, so it won't be supported which means that named decimal formats will not be available.

**Reply by ballsteve** [reply]
document() and xsl:number are now in the may-2024 branch

**Reply by ballsteve** [reply]
xsl:attribute-sets and xsl:use-attribute-sets are now in the may-2024 branch. It's a pretty awful implementation, but good enough for now (iow, I've never used them).

**Reply by ballsteve** [reply]
Abbreviated expressions in XPath are done. Extensions are optional, so I'm not going to worry about them for now.

So that's it! We have functional equivalence to XPath/XSLT 1.0 :-)

**Reply by Devasta** [reply]
Fantastic news, this is amazing work! Congratulations!

**Reply by ballsteve** [reply]
Thanks for your work too!

At this stage I'm going to merge into dev, and prepare for release to main and then publish.

**Reply by jonsterling** [reply]
This is really exciting! 🎆

**Comment by ballsteve** [comment]
Now that everything has ben pulled into the dev branch, I'm going to undertake the release engineering to publish a '1.0' version of xrust.
